core/src/main/java/org/bitcoinj/crypto/HDPath.java

```diff
@@ -0,0 +1,120 @@
+ * is used for rendering to String but (at present) not much else. It defaults to false which
+ * will also be the setting for a relative path.
+ *
+ * This class is immutable and uses the JDKs Collections.UnmodifiableList type internally.
+ *
+ * It implements {@code java.util.List<ChildNumber>} to ease migration
+ * from the previous Guava {@code ImmutableList<ChildNumber>}. It should be a minor breaking change
+ * to replace {@code ImmutableList<ChildNumber>} with {@code List<ChildNumber>} where necessary in your code. Although
+ * it is recommended to use the HDPath type for clarity and for access to HDPath-specific functionality.
+ */
+public class HDPath extends AbstractList<ChildNumber> {
+    static final char PREFIX_PRIVATE = 'm';
+    static final char PREFIX_PUBLIC = 'M';
+    static final char SEPARATOR = '/';
+    protected final boolean hasPrivateKey;
+    protected final List<ChildNumber> unmodifiableList;
+
+    public HDPath(boolean hasPrivateKey, List<ChildNumber> list) {
+        this.hasPrivateKey = hasPrivateKey;
+        this.unmodifiableList = Collections.unmodifiableList(list);
+    }
+
+    public HDPath(List<ChildNumber> list) {
+        this(false, list);
+    }
+
+    static HDPath of(boolean hasPrivateKey, List<ChildNumber> list) {
+        return new HDPath(hasPrivateKey, list);
+    }
+
+    static public HDPath of(List<ChildNumber> list) {
+        return HDPath.of(false, list);
+    }
+
+    static public HDPath of(ChildNumber childNumber) {
+        return HDPath.of(Collections.singletonList(childNumber));
+    }
+
+    static public HDPath of() {
+        return HDPath.of(Collections.<ChildNumber>emptyList());
+    }
+
+    /**
+     * Is this a path to a private key
+     * @return true if yes, false if no or a partial path
+     */
+    public boolean hasPrivateKey() {
+        return hasPrivateKey;
+    }
+
+    /**
+     * Extend the path by appending a ChildNumber
+     *
+     * @param child the child to append
+     * @return A new immutable path
+     */
+    public HDPath extend(ChildNumber child) {
+        List<ChildNumber> mutable = new ArrayList<>(this.unmodifiableList); // Mutable copy
+        mutable.add(child);
+        return new HDPath(mutable);
+    }
+
+    /**
+     * Extend the path by appending a relative path
+     *
+     * @param path2 the relative path to append
+     * @return A new immutable path
+     */
+    public HDPath extend(List<ChildNumber> path2) {
+        List<ChildNumber> mutable = new ArrayList<>(this.unmodifiableList); // Mutable copy
+        mutable.addAll(path2);
+        return new HDPath(mutable);
+    }
+
+
+    /**
+     * Extend the path by appending two ChildNumber objects
+     *
+     * @param child1 the first child to append
+     * @param child2 the second child to append
+     * @return A new immutable path
+     */
+    public HDPath extend(ChildNumber child1, ChildNumber child2) {
```

Comment:
nitpick: (style) maybe this should directly follow the `extend(ChildNumber child)` method.

Reply:
Fixed in force push: https://github.com/bitcoinj/bitcoinj/compare/7e832e4fd77a5cabd1df09caf3da946c4a140167..df2d543d4204193be42424bbaa6c15687cd82910